### PR TITLE
202406 native dimensionspecs

### PR DESCRIPTION
**Pull request by petermarshallio**
This notebook walks through some simple dimensionsSpec configurations - implicit, explicit, and autodiscovery.

### CODE REVIEW

File: notebooks/02-ingestion/15-native-dimensions.ipynb [hunk 1]
```diff
@@ -0,0 +1,786 @@
...
+    "\n",
+    "print(json.dumps(ingestionSpec, indent=5))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "fe50d1bb-1bc9-4ef1-a1a7-637cc3bb4616",
+   "metadata": {},
+   "source": [
+    "Start the ingestion of the raw data from Apache Kafka by submitting this object to Apache Druid by running the cell below."
```
Comment (ektravel) [inline notebooks/02-ingestion/15-native-dimensions.ipynb:394]:
```suggestion
    "Run the next cell to start ingestion raw data from Kafka to Druid."
```

File: notebooks/02-ingestion/15-native-dimensions.ipynb [hunk 2]
```diff
@@ -0,0 +1,786 @@
...
+    "supervisor = requests.post(f\"{druid_host}/druid/indexer/v1/supervisor\", json.dumps(ingestionSpec), headers=generalHeaders)\n",
+    "print(supervisor.status_code)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "dc897578-48d7-4b8c-9ece-dab4390e2336",
+   "metadata": {},
+   "source": [
+    "Run the following cell to wait until the ingestion has started and the new table is ready for query."
```
Comment (ektravel) [inline notebooks/02-ingestion/15-native-dimensions.ipynb:413]: "Run the cell to wait" sounds a bit odd.

File: notebooks/02-ingestion/15-native-dimensions.ipynb [hunk 3]
```diff
@@ -0,0 +1,786 @@
...
+    "druid.sql.wait_until_ready(table_name, verify_load_status=False)\n",
+    "print(\"Ready to go!\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "01430f85-eb06-4999-92f2-b789b3a4f611",
+   "metadata": {},
+   "source": [
+    "Run the following cell to get detail about the table you have created."
```
Comment (ektravel) [inline notebooks/02-ingestion/15-native-dimensions.ipynb:432]:
```suggestion
    "Run the following cell to get details about the table you have created."
```

File: notebooks/02-ingestion/15-native-dimensions.ipynb [hunk 4]
```diff
@@ -0,0 +1,786 @@
...
+  {
+   "cell_type": "markdown",
+   "id": "2654936f-8288-4fd5-a4ad-1f94267e43ae",
+   "metadata": {},
+   "source": [
+    "The `type` tells you how Druid will interpret the data in SQL. Notice that the `dimensionsSpec` has caused Druid to apply a type of BIGINT to `views`, `upvotes`, and `comments`.\n",
+    "\n",
+    "Learn more about data types in the dedicated [notebook on data types](../02-ingestion/04-table-datatypes.ipynb).\n",
+    "\n",
+    "Stop ingestion and drop the table programmatically by running the next cell."
```
Comment (ektravel) [inline notebooks/02-ingestion/15-native-dimensions.ipynb:454]:
```suggestion
    "Run the next cell to stop ingestion and drop the table."
```

File: notebooks/02-ingestion/15-native-dimensions.ipynb [hunk 5]
```diff
@@ -0,0 +1,786 @@
...
+    "druid.sql.wait_until_ready(table_name, verify_load_status=False)\n",
+    "print(\"Ready to go!\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "76e66285-c411-442e-9bdf-0b53e3322e57",
+   "metadata": {},
+   "source": [
+    "Review the schema for the table by running the next cell."
```
Comment (ektravel) [inline notebooks/02-ingestion/15-native-dimensions.ipynb:566]:
```suggestion
    "Run the next cell to view the schema for the table."
```

File: notebooks/02-ingestion/15-native-dimensions.ipynb [hunk 6]
```diff
@@ -0,0 +1,786 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "44738d6d-cec2-40ad-aaba-998c758c63f4",
+   "metadata": {},
+   "source": [
+    "## Clean up\n",
+    "\n",
+    "Run the following cell to stop the data generator."
```
Comment (ektravel) [inline notebooks/02-ingestion/15-native-dimensions.ipynb:728]: I recommend picking one type of intro and sticking with it (if possible). 
So either "Run the next cell to..." or "...by running the next cell."

File: notebooks/02-ingestion/15-native-dimensions.ipynb [hunk 7]
```diff
@@ -0,0 +1,786 @@
...
+   "source": [
+    "## Summary\n",
+    "\n",
+    "* The schema of incoming data is defined in the `dimensionsSpec` and is realized in the target table.\n",
+    "* Dimensions can be explicitly included and typed, explicitly excluded, or automatically detected and typed.\n",
+    "\n",
+    "## Learn more\n",
+    "\n",
+    "* Review the documentation on the [`dimensionsSpec`](https://druid.apache.org/docs/latest/ingestion/ingestion-spec#dimensionsspec).\n",
+    "* Read the official documentation on [partitioning](https://druid.apache.org/docs/latest/ingestion/partitioning) and [segment size optimization](https://druid.apache.org/docs/latest/operations/segment-optimization).\n",
```
Comment (ektravel) [inline notebooks/02-ingestion/15-native-dimensions.ipynb:754]:
```suggestion
    "* Review the documentation on [partitioning](https://druid.apache.org/docs/latest/ingestion/partitioning) and [segment size optimization](https://druid.apache.org/docs/latest/operations/segment-optimization).\n",
```

File: notebooks/02-ingestion/15-native-dimensions.ipynb [hunk 8]
```diff
@@ -0,0 +1,786 @@
...
+    "* Run through the dedicated [notebook on data types](../02-ingestion/04-table-datatypes.ipynb).\n",
+    "* Find out [how schemas can be changed](https://druid.apache.org/docs/latest/data-management/schema-changes) in Druid in the documentation.\n",
```
Comment (ektravel) [inline notebooks/02-ingestion/15-native-dimensions.ipynb:756]:
```suggestion
    "* Learn about [changing schemas](https://druid.apache.org/docs/latest/data-management/schema-changes) in Druid.\n",
```

File: notebooks/02-ingestion/15-native-dimensions.ipynb [hunk 9]
```diff
@@ -0,0 +1,786 @@
...
+    "* Experiment with combining batch and streaming data into the same table."
```
Comment (ektravel) [inline notebooks/02-ingestion/15-native-dimensions.ipynb:757]:
```suggestion
    "* Experiment with combining batch and streaming data in the same table."
```

**Review by ektravel — COMMENTED** [review]
Left some suggestions.

File: notebooks/02-ingestion/15-native-dimensions.ipynb [hunk 10]
```diff
@@ -0,0 +1,786 @@
...
+    "  ~\n",
+    "  ~ Unless required by applicable law or agreed to in writing,\n",
+    "  ~ software distributed under the License is distributed on an\n",
+    "  ~ \"AS IS\" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY\n",
+    "  ~ KIND, either express or implied.  See the License for the\n",
+    "  ~ specific language governing permissions and limitations\n",
+    "  ~ under the License.\n",
+    "  -->\n",
+    "\n",
+    "During streaming ingestion, the schema for events written into a table from a stream are set in the `dimensionsSpec`. This tutorial demonstrates various ways to work with the [dimensionsSpec](https://druid.apache.org/docs/latest/ingestion/ingestion-spec#dimensionsspec) against an example stream of events.\n",
```
[the remark this answers is not in the document]
Comment (petermarshallio) [reply notebooks/02-ingestion/15-native-dimensions.ipynb:28]: I've tried a new intro on this - we may need some wordsmithing later...

File: notebooks/02-ingestion/15-native-dimensions.ipynb [hunk 11]
```diff
@@ -0,0 +1,786 @@
...
+    "\n",
+    "tuningConfig = { \"type\": \"kafka\" }"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "d7ee01e6-d620-4ca7-9f2b-d49ab839ca5a",
+   "metadata": {},
+   "source": [
+    "The third part of the ingestion specification defines the [dataSchema](https://druid.apache.org/docs/latest/ingestion/ingestion-spec#dataschema). In the cells that follow, you will define all three parts:\n",
```
[the remark this answers is not in the document]
Comment (petermarshallio) [reply notebooks/02-ingestion/15-native-dimensions.ipynb:208]: Moved some stuff around... hoping it works better!

File: notebooks/02-ingestion/15-native-dimensions.ipynb [hunk 12]
```diff
@@ -0,0 +1,786 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "02beae03-25e9-4ce0-a785-ccfb13ec36cb",
+   "metadata": {},
+   "source": [
+    "## Explicitly set dimensions\n",
+    "\n",
+    "Run the next cell to create a `dimensionsSpec` object that uses the \"explicit\" method for ingesting events.\n",
```
[the remark this answers is not in the document]
Comment (petermarshallio) [reply notebooks/02-ingestion/15-native-dimensions.ipynb:317]: Added a little TOC above this.

File: notebooks/02-ingestion/15-native-dimensions.ipynb [hunk 13]
```diff
@@ -0,0 +1,786 @@
...
+    "\n",
+    "Notice that it is made up of [dimension objects](https://druid.apache.org/docs/latest/ingestion/ingestion-spec#dimension-objects) inside a `dimensions` list - the \"edited\" field has been left out intentionally.\n",
```
[the remark this answers is not in the document]
Comment (petermarshallio) [reply notebooks/02-ingestion/15-native-dimensions.ipynb:319]: AHA! Tis an array. Thanks :)

File: notebooks/02-ingestion/15-native-dimensions.ipynb [hunk 14]
```diff
@@ -0,0 +1,786 @@
...
+    "\n",
+    "There are two flavors of dimension object:\n",
```
[the remark this answers is not in the document]
Comment (petermarshallio) [reply notebooks/02-ingestion/15-native-dimensions.ipynb:321]: yeah.... I've removed it...

File: notebooks/02-ingestion/15-native-dimensions.ipynb (hunk 2 above)
Comment (petermarshallio) [reply notebooks/02-ingestion/15-native-dimensions.ipynb:413]: Yeah this was too lengthy... I've merged it into a single cell.

**Pull request merged**